rewrite based on LESS

braver opened the pull request:
Takes the LESS syntax as a basis for a complete rewrite of the SCSS syntax. Structurally the languages are very similar, with similar challenges.

The LESS syntax was [rewritten recently](https://github.com/danro/LESS-sublime/pull/79) over several months of testing and fine-tuning. It bases as much as possible on the default CSS package which has seen a lot of improvement since being open-sourced. Like the new default syntaxes LESS adheres to the [new scope naming guidelines](https://www.sublimetext.com/docs/3/scope_naming.html), and has similar nuances like scoping the `$` and `@` in variables and at-rules, or the quotes around string separately.

This also enables the completions from the default CSS package, fixing issues like #72 #93 #49.

This PR also fixes #85 and probably some other things too.

Let me know what you think. I also want to update the tests using Sublime's new test engine, and finally also upgrade the sass syntax.

braver committed:
rewrite based on LESS

braver committed:
add basic for/each

levymetal commented:
This is really good. I've had a quick play and I love the direction it's going, getting the autocomplete back would be amazing. 

Understandably the syntax highlighting is a bit off as this is currently a WIP, but it would be good to know what you're planning to do here.

In my opinion, the highlighting in the stable Sass package is _almost_ perfect, apart from a few small issues (line numbers refer to syntax_test.SCSS):

- highlighting for pseudo-classes and pseudo-elements (line 34)
- highlighting for css function names (`url()` etc) (line 203)
- improved highlighting within css functions in regards to variables, interpolation etc (line 235)

Your PR actually fixes those issues, but then changes a lot of the other highlighting. But I'm just not sure if this is because it's a WIP or if you've actively decided to change it. So before I bombard you with suggestions I'm keen to know what your plans are.

braver commented:
Thanks, glad you like it. Some notes:

Inevitably a lot of highlighting/colors will change, because it will match CSS wherever possible. It’s not feasible to try and keep the scopes in line with the current state of the package, and I think it’s a good idea to try to match the default languages. So, I hope we can look at the new highlighting and try to improve it, instead of looking at the delta with the current state and trying to minimize it. I definitely want your feedback for whatever you think should change though.

Another big change that might not be clear immediately is that it will use keyword lists for things like properties and values. This will help you catch typos like “colour” etc. This requires a little more maintenance, but if CSS, LESS and SASS align here it’s a lot easier to do.

Also, just so that’s clear, the package has some seemingly small issues as you say, but they cannot be fixed with small solutions. It needs big changes in how it decides something is a property or a selector (which causes the pseudo-element problem). The scopes are also way too loosely nested to ever be able to support auto completions. That’s the major reason for this PR: it’s the same problems LESS had going from a small naive syntax to one that’s robust and fully featured.

levymetal commented:
Ah I see, that makes sense. With that in mind, I've made a reduced test case [here](https://gist.github.com/levymetal/d978a9bea4ec363c54014c7331f25608) and broken down the issues I've noticed so far into important/not important categories.

There might also be other issues; `syntax_test.SCSS` is missing quite a few features from [SASS_REFERENCE](http://sass-lang.com/documentation/file.SASS_REFERENCE.html). I could put together a more comprehensive test file with all sass features if you think it would be useful - let me know.

Anyway, here are my thoughts so far:

**Important:**
- placeholder selectors (`%test`) should be highlighted like other selectors
- interpolation hash & curlies should be highlighted (`#{}`). this is especially important when used in a selector so the `#` doesn't get confused for an ID
- `@extend`, `@while`, `@error`, `@at-root`, `@debug`, `@warn`, `@return` should be highlighted

**Not so important:**
- tag selectors within pseudo-selectors should be highlighted like a tag (`:not(div)`)
- user-defined functions should be highlighted (`margin: test(4)`)

For reference, here are screenshots showing the current highlighting and the new highlighting using the [One Dark Theme](https://packagecontrol.io/packages/Theme%20-%20One%20Dark).

Current:
![old](https://user-images.githubusercontent.com/646278/39618013-3d5663b4-4fc5-11e8-82da-dda5624708a3.png)

New:
![new](https://user-images.githubusercontent.com/646278/39618021-43cb3e68-4fc5-11e8-94bc-d522b71ec203.png)

braver commented:
Thanks, that helps a lot. 

Looking at colors with just any color scheme is misleading though. It's not so much a question of "this should be highlighted", but rather "this should be scoped correctly". A lot of color schemes aren't ready for the scopes used in new Sublime Text 3 syntaxes. We can try to be a bit "backwards compatible" here, but it shouldn't be the primary target because someone else can come around with a favourite color scheme and it may or may not get the "right" colors there. If you're going to test, the Mariana color scheme is a good one to use.

levymetal commented:
Thanks for the info. You're right, I'm describing the result not the mechanism. In my head I'm thinking in terms of _highlighting_, not _scopes_; I've been writing CSS for 10 years but never considered what happens behind the scenes in the editor.

I've already learnt a lot today, I'll continue to put effort into learning more about scopes. Hopefully this will enable me to better understand the issues you're facing and provide relevant feedback about how I think the scopes should be applied as opposed to my current feedback of "make it look like this" 😂.

braver commented:
Think of it like this:

- The color scheme holds the CSS. It tells the editor to make `support.function` blue (or whatever).
- The syntax defines what classes to put on the "HTML". It tells the editor that `function()` is a `support.function`, so that the color scheme can apply the color.

So, we're trying to get the semantics right here, not the colors. Of course, if we radically change the semantics, and the semantics in most Sublime Text 3 syntaxes have gotten a lot more detailed, old color schemes may not have colors for them. Right now the default color schemes are really excellent, but most of the old favourites are years old, so there is only so much we can do there.

braver committed:
move the tests

braver committed:
revert the parameters

braver committed:
tweak placeholder selector

braver committed:
rename

braver committed:
update property names list

braver committed:
fix tag selector inside :not()

braver commented:
Tackled: 

> tag selectors within pseudo-selectors should be highlighted like a tag (:not(div))

Apparently a bug in LESS as well, so win-win there. 

Also:
> placeholder selectors (%test) should be highlighted like other selectors

This should already be good, but probably not supported by your color scheme. Looks as expected in Mariana though:
>interpolation hash & curlies should be highlighted (#{}). this is especially important when used in a selector so the # doesn't get confused for an ID

I'll add more tests from the reference docs next week and hopefully tackle some more issues.

I'll also try and update some things based on improvements in the default CSS package. Because of the nested nature of LESS/SCSS and things like interpolation it's not just a copy-paste thing, but it's set up to not be that much effort either.

levymetal commented:
Nice. I've switched to Mariana, and with these latest updates, it's looking great. I'm now using this PR instead of the stable package as the improvements are well worth it.

Also, I've edited my previous comment to remove the irrelevant issues that were caused by my old color scheme. I absolutely agree with your statement about using the new ST3 scopes and limiting backwards compatibility. It's the color scheme's job to support new syntax scopes, not the syntax's job to support old color schemes.

braver committed:
add more css functions from default CSS package

braver committed:
add nested properties, start building tests

braver committed:
fix comment toggling

braver committed:
add test for variables and default|global keywords

braver committed:
move comments tests

braver committed:
add tests for mixin, strings, interpolation

braver committed:
add tests for {}

braver committed:
sort the includes

braver committed:
add tests for strings

braver committed:
add tests and make fixes: move & to selector

braver committed:
add lists and maps

braver commented:
@P233 Are you at all interested in me pursuing this further? Have any feedback? I'm getting a lot of value out of this also for the LESS package, developing both side-by-side has a lot of synergy. I also wouldn't mind going through some of the open issues here, see if I can either fix something in this PR or otherwise help people out.

braver committed:
add all functions and test for color functions

braver committed:
(unquoted) strings and interpolation

braver committed:
special unquoted string concats in property values

braver committed:
docs moved to https

braver committed:
tweak unquoted string concats

braver committed:
negative check on bs string property value

braver committed:
simpler, more robust string concatenation

braver committed:
fix chained selector by updating from CSS

ignoring custom-elements for now

braver committed:
make sure selector != value or property

braver committed:
() = punctuation.definition, not .section

braver committed:
reduce operators to actual scss spec

braver committed:
add some basic ignores

braver committed:
improve selectors and combinators

michaelblyons commented on Syntaxes/SCSS.sublime-syntax:
```diff
@@ -230,12 +231,17 @@ contexts:
 
   property-list:
-    - match: '(!#)*\{'
+    - match: '(\s*)\{'
```
You may want to change this to `\s*(\{)` and change the `scope` to a `capture` on group 1. Otherwise, you are scoping the white space as `punctuation`.

braver replied:
Thanks!

braver commented:
Ping @P233 any opinions about this?

olivermuc commented:
It doesn't look like it's being maintained anymore, can you fork? @braver

braver commented:
@olivermuc if @P233 doesn't respond I can publish my fork in its place. I still need some time for tests and iteration, so not necessarily in a hurry. If you want to use this right now you can manually install my form from https://github.com/braver/Syntax-highlighting-for-Sass

olivermuc commented:
@braver sounds great. btw. I just pulled your request locally and tried it, but must be missing something. ie. changing type/scope to scss or sass is not respecting the css autocomplete part. i may just be dense here.
Edit: I'll try your GIT repo - not sure if this makes any difference, though.

braver commented:
@olivermuc make sure you're on my `rewrite` branch. You should get css auto-completions:

<img width="525" alt="screen shot 2018-06-18 at 15 36 17" src="https://user-images.githubusercontent.com/2543659/41539189-6e20fde0-730d-11e8-9582-713456d486e9.png">

I haven't pushed my latest changes, so that might be a problem too. I'll review and push them ASAP.

olivermuc commented:
Right. Just tried your own repo and same behavior, I can switch to SCSS, but don't get the CSS3 autocomplete popup.
EDIT: 3rd attempt resulted in SCSS to now show the CSS autocomplete popup, however, I lost all of the SASS features.

braver committed:
() = punctuation.definition

braver committed:
reduce operators

braver committed:
remove custom element stuff

braver committed:
improve selectors

braver committed:
add tests

braver committed:
Merge commit '1a8a52d0e8c27162fe63dc3225bf19da78bb760d' into rewrite

# Conflicts:
#	Tests/syntax_test_scss.SCSS

braver commented:
@olivermuc I pushed some stuff, does it work now?

olivermuc commented:
Unfortunately no difference, with type SCSS selected, I get the css3 autocomplete, but not the SASS features (eg. variable autocomplete, color functions)

olivermuc commented:
Do I need to change any settings?

braver commented:
@olivermuc yes, that's correct. I've been meaning to try and and get both. Sublime should pick up on the variables by itself (screenshot), but the color functions have to be provided by the plugin. 

<img width="263" alt="screen shot 2018-06-18 at 15 52 53" src="https://user-images.githubusercontent.com/2543659/41540204-e51e4bd0-730f-11e8-9d6d-514914d00fe1.png">

olivermuc commented:
I removed both CSS3 and yours, reinstalled afresh, but same as before: css autocomplete, but no variable autocomplete (or color functions or any SASS stuff really).

EDIT: not using CSS3 but the standard CSS support does allow me to auto complete variables.
Just not any other SASS function.

braver commented:
For variables you need to type `$` and the first character of the variable. As I mentioned, there is no auto-complete for SASS features. 

This package, without my PR, also does not offer those features though.

olivermuc commented:
Apologies, must have missed that part. All good then, Koen - thanks for your prompt support!
Highly appreciated.

braver committed:
add snippets for sass functions

https://sass-lang.com/documentation/Sass/Script/Functions.html
Snippets, not completions, because that interferes with CSS completions

braver commented:
@olivermuc I've added a bunch of completions for SASS functions.

olivermuc commented:
@braver You are a star! Works like a charm and is helping tons.

braver committed:
add the sublime-completions to git

braver committed:
remove sublime-completions because of Core #819

https://github.com/SublimeTextIssues/Core/issues/819

pratikborsadiya commented:
I wanted this for ages. Thank You @braver for properly enabling syntax highlighting and autocomplete for SASS.

braver committed:
transparentize separate second marker

andremacola commented:
@braver I get no variables autocomplete. Everything else is working.

What I'm possibly doing wrong? Using the rewrite branch

![captura de tela 2018-07-11 as 22 13 57](https://user-images.githubusercontent.com/3408809/42607150-d8f66950-8557-11e8-9d38-e810568aa197.png)

braver commented:
The properties and snippets take precedence, and there’s a bunch of them here. It also looks like you’re using Emmet, which introduces even more completions to that list.

PS. I should be able to make variable autocomplete a bit better, not sure how that works yet.

braver committed:
fix commenting indented lines by overruling CSS

braver committed:
stop inserting ; in selectors (e.g. a:hover)

andremacola commented:
@braver for some reason, variables auto complete is not working with me.
another problem is with the syntax highlight. See the image below. Problem with `()`

![captura de tela 2018-07-27 as 03 29 03](https://user-images.githubusercontent.com/3408809/43305231-509f8102-914d-11e8-9718-e0c98cfd17a9.png)

braver committed:
fix nested grouping

braver commented:
@andremacola thanks for reporting this, I missed this case. Should be fixed now.

What exactly are you doing when you expect variables to appear in the auto-complete? I'm not facilitating anything special here yet, so `$s` should list word ST found starting with `s`, but it's not really "variables autocomplete" specifically (no language in ST has anything more special here by the way, not without some codeintel or language server running in the background).

braver commented:
@P233 do you mind if I replace the entry for this package in package control with my fork? Hate to put an ultimatum on things, but at some it has to go one way or the other.

braver committed:
normalize extensions

braver committed:
remove old tests

braver committed:
align the sass (indented) syntax

braver committed:
minor cleanup

braver committed:
update readmes

ghost commented:
@braver Amazing work! Any hopes making this work with Vue.js single file components? P233’s original code is working, my guess is it breaks with the updated scopes. Cheers!

braver commented:
@indrekpaas I believe the vue syntax package was at some point updated to support the [new scopes for LESS](https://github.com/vuejs/vue-syntax-highlight/blob/new/Vue%20Component.sublime-syntax#L783), the same thing could be done for SCSS. However, since @P233 decided against merging this PR and instead might at some point perhaps hopefully update the syntax himself, there will be for the foreseeable two syntax packages for SCSS/Sass (well, three if you count the even more broken and outdated textmate package). 

It is not a good outcome, but it's @P233's repo, so that's what we've got to work with.

So, the Vue syntax will have to decide to support this one, or my fork. Or you could fork the Vue syntax package and make it use the [scopes](https://github.com/vuejs/vue-syntax-highlight/blob/new/Vue%20Component.sublime-syntax#L727) that are supported by my fork. I can't support the old scopes without losing support for CSS autocompletions, or having to rebuild and then maintain those completions myself. That is an option of course, the CSS and SCSS/LESS syntaxes aren't fully compatible and things like semi-colon insertion are a bit tricky, so I might at some point have to do this anyway. I'd rather not.

My fork should become available on Package Control soon by the way.

Thom1729 commented:
@P233 Are you still interested in maintaining this package? If you aren't anymore, it would be possible to transition the package to @braver's repository, which would be preferable to having an extra package around. If you still want to maintain this one and just take it in another direction, that's fine too.

FichteFoll commented:
As a long-time maintainer of the Package Control channel, I would also like to voice my support for having this be integrated into the **very** popular `Sass` package so that many people may benefit from these improvements rather than only those few that happen to come along [the rewrite](https://github.com/wbond/package_control_channel/pull/7220).
